modules/fedimint-ln/src/contracts/mod.rs

```diff
@@ -112,3 +113,76 @@ impl Decodable for ContractId {
         Ok(ContractId::from_inner(Decodable::consensus_decode(d)?))
     }
 }
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize, Serialize, Encodable, Decodable)]
+pub struct Preimage(pub [u8; 32]);
+
+impl Preimage {
+    /// Create a Schnorr public key from this preimage
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Error::InvalidPublicKey`] if the Preimage does not represent a valid Secp256k1 point x coordinate.
+    pub fn to_public_key(&self) -> Result<secp256k1::XOnlyPublicKey, secp256k1::Error> {
+        secp256k1::XOnlyPublicKey::from_slice(&self.0)
+    }
+}
+
+/// Possible outcomes of preimage decryption
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize, Serialize, Encodable, Decodable)]
+pub enum DecryptedPreimage {
+    /// There aren't enough decryption shares yet
+    Pending,
+    /// The decrypted preimage was valid
+    Some(Preimage),
+    /// The decrypted preimage was invalid
+    Invalid,
+}
+
+/// Threshold-encrypted [`Preimage`]
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize, Serialize)]
+pub struct EncryptedPreimage(pub threshold_crypto::Ciphertext);
+
+/// Share to decrypt an [`EncryptedPreimage`]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct PreimageDecryptionShare(pub threshold_crypto::DecryptionShare);
+
+impl EncryptedPreimage {
+    pub fn new(preimage: Preimage, key: &threshold_crypto::PublicKey) -> EncryptedPreimage {
+        EncryptedPreimage(key.encrypt(preimage.0))
+    }
+}
+
+impl Encodable for EncryptedPreimage {
+    fn consensus_encode<W: std::io::Write>(&self, writer: W) -> Result<usize, Error> {
+        // TODO: get rid of bincode
+        let bytes = bincode::serialize(&self.0).expect("Serialization shouldn't fail");
+        bytes.consensus_encode(writer)
+    }
+}
+
+impl Decodable for EncryptedPreimage {
+    fn consensus_decode<D: std::io::Read>(d: D) -> Result<Self, DecodeError> {
+        let bytes = Vec::<u8>::consensus_decode(d)?;
+        Ok(EncryptedPreimage(
+            bincode::deserialize(&bytes).map_err(DecodeError::from_err)?,
+        ))
+    }
+}
+
+impl Encodable for PreimageDecryptionShare {
+    fn consensus_encode<W: std::io::Write>(&self, writer: W) -> Result<usize, Error> {
+        // TODO: get rid of bincode
+        let bytes = bincode::serialize(&self.0).expect("Serialization shouldn't fail");
+        bytes.consensus_encode(writer)
+    }
+}
+
+impl Decodable for PreimageDecryptionShare {
+    fn consensus_decode<D: std::io::Read>(d: D) -> Result<Self, DecodeError> {
+        let bytes = Vec::<u8>::consensus_decode(d)?;
+        Ok(PreimageDecryptionShare(
+            bincode::deserialize(&bytes).map_err(DecodeError::from_err)?,
+        ))
+    }
+}
```

Comment:
This stuff is very specific to the incoming contract.

Reply:
Not sure which stuff in particular. If you mean the whole lot related to Preimage Encryption/Decryption, I do agree the application is heavily skewed to incoming contract, though I'm not sure that the impls are imcomping contract specific